### DISCUSSION TITLE
How to run qutebrowser in headless kiosk mode with forced pass-through?

### DISCUSSION THREAD

**Discussion by arija-ari** [qutebrowser/qutebrowser#8429] [Q&A] [answered]
Hi,

I'm on MacOS (Intel) with an eGPU and I have to deal with a 3D web application (a game) that only runs exceptionally well under qutebrowser. I believe this is due to QtWebEngine and the fact that qutebrowser does not launch any subprocesses and thus can fully use my eGPU.

- WebKit (Safari) would maybe run better, if it would properly run on my eGPU, but it does not and this is Apple's fault and they confirmed this already
- Chrome-based browsers perform worse (this is due to poor programming of the game) and they have annoying GPU/Helper sub apps, that I have to tell to prefer my eGPU over and over again, because those settings will reset with every update
- Firefox performs much better, but something about the web application doesn't work well with the Firefox engine and FPS lag spikes occur every 2 seconds (this is also due to poor programming of the game)
- I cannot fix the game (proprietary), so I tried to circumvent the issue

By accident I discovered this browser and the QtWebEngine. Using this browser, the game runs as fluid as in Firefox, but without the lag spikes. Now, I want to make it work for my use-case, because any other known QtWebEngine browser alternative is either outdated or Linux-only.

1. How can I launch a new window of qutebrowser right into fullscreen
2. How can I launch such window in headless mode, so that it disables tabs and status bar
3. How can I force passthrough-mode on such window without the ability to exit
4. Optionally, is there a way to do it only via a command but still use the browser windowed in regular way, because now I am intrigue to try it out
5. Optionally, is there a way to control saturation of the rendered canvas, it seems the game looks much more saturated in qutebrowser only, but I'm willing to accept that

Thank you so much. ❤️

**Comment by arija-ari** [comment, marked as answer]
So it is all very possible with some minor caveats. I read myself into https://qutebrowser.org/doc/help/configuring.html#configpy so I created myself a Python file in `$HOME/.qutebrowser/config_arcade.py`:

```py
# pyright: reportUndefinedVariable=false
config.load_autoconfig(False)
c.statusbar.show = 'never'
c.tabs.show = 'never'
c.tabs.tabs_are_windows = True
c.window.hide_decoration = True
c.zoom.levels = ['100%']
c.zoom.text_only = True
config.unbind('<F11>', mode='normal')
config.bind('<Escape>', 'mode-enter passthrough')
config.unbind('<Shift-Escape>', mode='passthrough')
```

- It hides tab and status bar
- It prevents zooming
- It prevents changing fullscreen
- It enables entering passthrough through Escape and prevents exiting pass-through

Now to launch my game I do: `qutebrowser -C "$HOME/.qutebrowser/config_arcade.py" --target window ":fullscreen --enter" https://universe.flyff.com/play`

Two caveats:

1. In order to start interacting with the game, I need to hit Escape once to enter pass-through mode
2. #8430 is currently a thing that I couldn't solve, but I can just close the popup before entering pass-through and it will be alright

**Reply by The-Compiler** [reply]
1. You could probably set `input.mode_override` to `passthrough` (normally done per-domain). Or remove all bindings via `c.bindings.default = {}` and not use passthrough mode at all.
2. Perhaps you can pass `":open https://universe.flyff.com/play"` as a command instead, or an immediate `:tab-close` command (maybe with `:cmd-later`?)

I'm surprised it works so well for you while Chromium does not, given that it's also based on Chromium!

**Reply by arija-ari** [reply]
@The-Compiler I now believe I know why it runs good and Chromium does not. Today I got the new qutebrowser update 3.5.0 which uses the new QtWebEngine and suddenly the game runs poorly as if I was using Chrome. So I guess QtWebEngine was playing nice with an older Chromium upstream and some recent changes introduce problems with such game and Qt was just late on adopting changes. I know for a fact that only Intel devices are affected and that the game itself is at fault, other games work fine.

It's such an edge case, and most Macs are Apple Silicon, so I feel left alone, most gamers are Windows users, so I feel left alone, I'm a very unique case in the player base. I decided to lock qutebrowser's version on the latest 3.4 (I downgraded), so I can enjoy the game the way I do not for months (it actually worked to downgrade). I will hence not use it as a daily driver for anything else to prevent and security issues (the game is really just a canvas) and I trust the game provider.

What a bummer.

**Reply by mschilli87** [reply]
Have you tested if it works with newer versions of Chromium/QtWebengine under Linux?
Maybe upgrading your OS from MacOS to Linux would be an overall better solution than downgrading your browser. 😉 
Also, if you know the game is at fault and trust the dev, why not try to work with them on fixing your issue directly in the game?

**Reply by arija-ari** [reply]
@mschilli87 I have a Linux machine here, but it's not powerful enough to get an accurate comparison. My Mac is too full to dual boot.

I agree that it was better to move to Linux as my daily driver, but there are a lot of challenges I face and I still need access to Xcode introducing more challenges. A swap to different hardware seems a good idea, but I'm locked in due to convenience. It's hard to explain, but the built-in keyboard and touchpad provide me with value that no other laptop can give me. Maybe someday I have the patience to pull-through.

I tried to work with the devs to fix the game and I trust the devs, but I can't rely on their management. The lead dev is testing with Apple Silicon and their management puts priorities on other things. I'm part of a minority of a minority. Everyone tells me to get a gaming PC, but I refuse to do so, because I am a 10 finger typist (muscle memory hits home) and embrace simplicity (one device to rule them all). 😌
